v2/blockstore/writeonly.go

```diff
@@ -0,0 +1,187 @@
+package blockstore
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/ipfs/go-cid"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	format "github.com/ipfs/go-ipld-format"
+	blocks "github.com/ipfs/go-libipfs/blocks"
+	carv2 "github.com/ipld/go-car/v2"
+	"github.com/ipld/go-car/v2/internal/carv1"
+	"github.com/ipld/go-car/v2/internal/carv1/util"
+)
+
+var ErrWriteOnly = errors.New("write-only blockstore; unsupported operation")
+
+var _ blockstore.Blockstore = (*WriteOnly)(nil)
+
+// WriteOnly is a blockstore that only supports write operations, Put and
+// PutMany as well as some indexed operations: Has, GetSize and AllKeysChan.
+type WriteOnly struct {
+	mu     sync.RWMutex
+	opts   carv2.Options
+	out    io.Writer
+	blocks map[string]wroteBlock
+}
+
+type wroteBlock struct {
+	cid  cid.Cid
+	size int
+}
+
+// CreateWriteOnlyV1 creates a new WriteOnly blockstore that writes to the given
+// io.Writer in CARv1 format. A WriteOnly CARv1 is able to support most of the
+// Blockstore interface while also being suitable for writing to a streaming
+// output as it only requires an io.Writer.
+//
+// The returned WriteOnly blockstore will not support Get or DeleteBlock
+// operations, but keeps track of written CIDs so is able to support Has,
+// GetSize and AllKeysChan operations.
+func CreateWriteOnlyV1(out io.Writer, roots []cid.Cid, opts ...carv2.Option) (*WriteOnly, error) {
```

Comment:
```suggestion
func NewWriteOnlyV1(out io.Writer, roots []cid.Cid, opts ...carv2.Option) (*WriteOnly, error) {
```

I think this might be more idiomatic